(maint) Add Initial set templates (#1)
chocolatey/.github: CONTRIBUTING.md

      - [Submitting an Enhancement For Choco](#submitting-an-enhancement-for-choco)
  - [Contributing](#contributing-1)
    - [Prerequisites](#prerequisites)
      - [Definition of Trivial Contributions](#definition-of-trivial-contributions)
      - [Is the CLA Really Required?](#is-the-cla-really-required)
  - [Contributing Process](#contributing-process)
    - [Get Buyoff Or Find Open Community Issues/Features](#get-buyoff-or-find-open-community-issuesfeatures)
    - [Set Up Your Environment](#set-up-your-environment)
    - [Code Format / Design](#code-format--design)
      - [CSharp](#csharp)
      - [PowerShell](#powershell)
    - [Debugging / Testing](#debugging--testing)
      - [Visual Studio](#visual-studio)
        - [Automated Tests](#automated-tests)
      - [Chocolatey Build](#chocolatey-build)
    - [Prepare Commits](#prepare-commits)
    - [Submit Pull Request (PR)](#submit-pull-request-pr)
    - [Respond to Feedback on Pull Request](#respond-to-feedback-on-pull-request)
  - [Other General Information](#other-general-information)

<!-- /TOC -->

## Are You In the Right Place?

> Chocolatey is a large ecosystem and each component has their own location for submitting issues and enhancement requests. While the website (the community package repository) may be all you know for packages, it represents only a tiny fraction of existing packages (organizations typically maintain and host their own packages internally). This is the repository for choco.exe (the client CLI tool) for Chocolatey, which spans multiple types of environments.

Reviewer: Same as @JPRuskin noted, this needs to be made generic.

Please follow these decision criteria to see if you are in the right location or if you should head to a different location to submit your request.

### Reporting an Issue/Bug?
![submitting issues](https://cloud.githubusercontent.com/assets/63502/12534440/fc223b74-c21e-11e5-9a41-1ffc1c9af48f.png)

> Submitting an Issue (or a Bug)? See the **[Submitting Issues](https://github.com/chocolatey/choco#submitting-issues) section** in the [README](https://github.com/chocolatey/choco/blob/master/README.md#submitting-issues).

Reviewer: Again, should be made generic

Please see [Request Package Fixes or Updates / Become a maintainer of an existing package](https://docs.chocolatey.org/en-us/community-repository/users/package-triage-process).

### Package Request? Package Missing?
If you are looking for packages to be added to the community feed (aka https://community.chocolatey.org/packages), please see [Package Requests](https://docs.chocolatey.org/en-us/community-repository/users/package-triage-process#package-request-package-missing).

### Submitting an Enhancement / Feature Request?
If this is for choco (the CLI tool), this is the right place. See below. Otherwise see [Submitting Issues](https://github.com/chocolatey/choco#submitting-issues) for enhancements to the website, enhancements to the ChocolateyGUI, etc.

#### Submitting an Enhancement For Choco
Log a github issue. There are fewer constraints on this versus reporting issues.

## Contributing
The process for contributions is roughly as follows:

### Prerequisites
 * Submit the Enhancement ticket. You will need the issue id for your commits.
 * Ensure you have signed the Contributor License Agreement (CLA) - without this we are not able to take contributions that are not trivial.

>   * [Sign the Contributor License Agreement](https://cla-assistant.io/chocolatey/choco).

Reviewer 2: After speaking with @gep13 it appears we using it for Chocolatey GUI as well. I've removed the Chocolatey CLI specific link in this [commit](https://github.com/chocolatey/.github/pull/1/commits/24eaf85188d87bc54f3b75cff391c599663ff44c).

If you work for an organization that does not allow you to contribute without attempting to own the rights to your work, please do not sign the CLA.

## Contributing Process

Start with [Prerequisites](#prerequisites) and make sure you can sign the Contributor License Agreement (CLA).

### Get Buyoff Or Find Open Community Issues/Features
 * Through a Github issue (preferred), through the [mailing list](https://groups.google.com/forum/#!forum/chocolatey), or through [Gitter](https://gitter.im/chocolatey/choco), talk about a feature you would like to see (or a bug fix), and why it should be in Chocolatey.
   * If approved through the mailing list or in Gitter chat, ensure the accompanying GitHub issue is created with information and a link back to the mailing list discussion (or the Gitter conversation).
 * Once you get a nod from one of the [Chocolatey Team](https://github.com/chocolatey?tab=members), you can start on the feature.

>  * Alternatively, if a feature is on the issues list with the [Up For Grabs](https://github.com/chocolatey/choco/issues?q=is%3Aopen+is%3Aissue+label%3A%22Up+For+Grabs%22) label, it is open for a community member (contributor) to patch. You should comment that you are signing up for it on the issue so someone else doesn't also sign up for the work.

Reviewer 3: Again, somewhat unsure this will work.
```suggestion
 * Alternatively, if a feature is on the issues list with the [Up For Grabs](./issues?q=is%3Aopen+is%3Aissue+label%3A%22Up+For+Grabs%22) label, it is open for a community member (contributor) to patch. You should comment that you are signing up for it on the issue so someone else doesn't also sign up for the work.

```

### Set Up Your Environment
 * Visual Studio 2010+ is recommended for code contributions.
 * For git specific information:

>     1. Create a fork of chocolatey/choco under your GitHub account. See [forks](https://help.github.com/articles/working-with-forks/) for more information.

Reviewer 3: ```suggestion
    1. Create a fork of this repository under your GitHub account. See [forks](https://help.github.com/articles/working-with-forks/) for more information.

```

    1. [Clone your fork](https://help.github.com/articles/cloning-a-repository/) locally.
    1. Open a command line and navigate to that directory.

>     1. Add the upstream fork - `git remote add upstream git@github.com:chocolatey/choco.git`

Reviewer 3: ```suggestion
    1. Add the upstream fork - e.g. `git remote add upstream git@github.com:chocolatey/choco.git`

```

    1. Run `git fetch upstream`
    1. Ensure you have user name and email set appropriately to attribute your contributions - see [Name](https://help.github.com/articles/setting-your-username-in-git/) / [Email](https://help.github.com/articles/setting-your-email-in-git/).
    1. Ensure that the local repository has the following settings (without `--global`, these only apply to the *current* repository):

>       * `git config core.autocrlf false`

Reviewer: Someone should ensure that these are the correct settings to use for other repositories as well.

 * The minimum version of PowerShell this must work with is v2. This makes things somewhat more limited but compatible across the board for all areas Chocolatey is deployed. It is getting harder to find a reference for PowerShell v2, but this is a good one: http://adamringenberg.com/powershell2/table-of-contents/.
 * If you add a new file, also ensure you add it to the Visual Studio project and ensure it becomes an embedded resource.
 * The last parameter in every function must be `[parameter(ValueFromRemainingArguments = $true)][Object[]] $ignoredArguments`. This allows for future expansion and compatibility - as new parameters are introduced and used, it doesn't break older versions of Chocolatey.
 * Do not add new positional elements to functions. We want to promote using named parameters in calling functions.
 * Do not remove any existing positional elements from functions. We need to maintain compatibility with older versions of Chocolatey.
 * One of the first calls in a function is to debug what was passed to it - `Write-FunctionCallLogMessage -Invocation $MyInvocation -Parameters $PSBoundParameters`

### Debugging / Testing

> When you want to manually verify your changes and run Choco, you have some options.

Reviewer 3: ```suggestion
When you want to manually verify your changes, you have some options.
```

> **NOTE:** Chocolatey behaves differently when built with `Debug` and `Release` configurations. Release is always going to seek out the machine installation (`$env:ChocolateyInstall`), where Debug just runs right next to wherever the choco.exe file is.

Reviewer 3: ```suggestion
```

 * Blackbox Testing - tests where you do not access internals of the application
    * Physical Integration Testing - This is where you are testing the application with other components such as config files.
    * Blackbox Integration Testing / End to End Testing - This is where you are testing inputs and outputs of the system.

As far as testing goes, unit tests are extremely quick feedback and great for longer term maintenance, where black box tests give you the most coverage, but are the slowest feedback loops and typically the most frail. Each area of testing has strengths and weaknesses and it's good to understand each of them.

**NOTE**: One of the hardest forms of automated testing is unit testing, as it almost always requires faking out other parts of the system (also known as mocking).

> #### Chocolatey Build

Reviewer: Maybe the choco cli specific items could be factored out into a separate document (CONTRIB_PER_REPO.md), and linked to from here? And other repository-specific items in other repositories could be added.
